agent/logs.go

```diff
@@ -0,0 +1,172 @@
+		l.queues[src] = q
+	}
+	for _, log := range logs {
+		pl, err := agentsdk.ProtoFromLog(log)
+		if err != nil {
+			return xerrors.Errorf("failed to convert log: %w", err)
+		}
+		q.logs = append(q.logs, pl)
+	}
+	logger.Debug(context.Background(), "enqueued agent logs", slog.F("new_logs", len(logs)), slog.F("queued_logs", len(q.logs)))
+	return nil
+}
+
+func (l *logSender) flush(src uuid.UUID) error {
+	l.L.Lock()
+	defer l.L.Unlock()
+	defer l.Broadcast()
+	q, ok := l.queues[src]
+	if ok {
+		q.flushRequested = true
+	}
+	// queue might not exist because it's already been flushed and removed from
+	// the map.
+	return nil
+}
+
+// sendLoop sends any pending logs until it hits an error or the context is canceled.  It does not
+// retry as it is expected that a higher layer retries establishing connection to the agent API and
+// calls sendLoop again.
+func (l *logSender) sendLoop(ctx context.Context, dest logDest) error {
+	ctxDone := false
+	defer l.logger.Debug(ctx, "sendLoop exiting")
+
+	// wake 4 times per flush interval to check if anything needs to be flushed
+	go func() {
+		tkr := time.NewTicker(flushInterval / 4)
+		defer tkr.Stop()
+		for {
+			select {
+			// also monitor the context here, so we notice immediately, rather
+			// than waiting for the next tick or logs
+			case <-ctx.Done():
+				l.L.Lock()
+				ctxDone = true
+				l.L.Unlock()
+				l.Broadcast()
+				return
+			case <-tkr.C:
+				l.Broadcast()
+			}
+		}
+	}()
+
+	l.L.Lock()
+	defer l.L.Unlock()
+	for {
+		for !ctxDone && !l.hasPendingWorkLocked() {
+			l.Wait()
+		}
+		if ctxDone {
+			return nil
+		}
+		src, q := l.getPendingWorkLocked()
+		q.flushRequested = false // clear flag since we're now flushing
+		req := &proto.BatchCreateLogsRequest{
+			LogSourceId: src[:],
+			// when we release the lock, we don't want modifications to the slice to affect us
+			Logs: slices.Clone(q.logs),
+		}
+
+		l.L.Unlock()
+		l.logger.Debug(ctx, "sending logs to agent API", slog.F("log_source_id", src), slog.F("num_logs", len(req.Logs)))
+		_, err := dest.BatchCreateLogs(ctx, req)
+		l.L.Lock()
+		if err != nil {
+			return xerrors.Errorf("failed to upload logs: %w", err)
+		}
+
+		// since elsewhere we only append to the logs, here we can remove them
+		// since we successfully sent them
+		q.logs = q.logs[len(req.Logs):]
+		if len(q.logs) == 0 {
+			// no empty queues
+			delete(l.queues, src)
+			continue
+		}
+		q.lastFlush = time.Now()
+	}
+}
+
+func (l *logSender) hasPendingWorkLocked() bool {
+	for _, q := range l.queues {
+		if q.flushRequested {
+			return true
+		}
+		if time.Since(q.lastFlush) > flushInterval {
+			return true
+		}
+	}
+	return false
+}
+
+func (l *logSender) getPendingWorkLocked() (src uuid.UUID, q *logQueue) {
+	// take the one it's been the longest since we've flushed, so that we have some sense of
+	// fairness across sources
+	var earliestFlush time.Time
+	for is, iq := range l.queues {
+		if q == nil || iq.lastFlush.Before(earliestFlush) {
```

Comment:
We don't allow zero-length queues.

I mostly like the implementation of just looking at the earliest flush because it's easy to understand and get correct.

While it's true that this implementation might flush another queue before the one that has `flushRequested`, I think it's desirable for there to be some fairness among sources, so a source that flushes after every log can't just hog the RPC.

Right now `flushRequested` doesn't allow you to jump the queue, but it means we don't "rest" until the queue that requested it has been flushed, which seems OK to me.

Reply:
Yeah, I think it's fine. Considering flushRequested isn't prioritized, though, I might change the order of the `if` checks in `hasPendingWorkLocked`, just so that it doesn't lead anyone astray. 😄